### ISSUE TITLE
ARAP python implementation

### ISSUE THREAD

**Issue by GiBucci** [riccardomarin/FARM#3]
Hi.

First of all thank you for sharing your work, it is great!

I am trying to achieve the same thing you did, only I already have an implementation of shape registration with the SMPL template.

I would love to apply fine tuning to the output of the model, since it is not as precise as I would love it to be.
I found your ARAP algorithm and it seems it is great for what I am trying to achieve.

The only problem is I would like to have a pure python implementation.
I am trying to understand what you did but I am not so skilled in Matlab, neither I am in geometry so it is very hard for me to "translate" the process.

In conclusion what I would like to know is:
- do you already have a pure python implementation of the last step as a continuation of this work? Or is it planned in the future?
- If not, can you point me to some resource where it is tackled in the same way as you did? I found many implementations of this apporach for deformation with handles, but none about shape registration.

Thank you in advance for time and consideration.

Have a nice day.

**Comment by riccardomarin** [comment]
Hi Giovanni!
Thank you for your interest in our work!

The ARAP energy we used comes from the Matlab GPToolBox: https://github.com/alecjacobson/gptoolbox
In our work we proceeded as follows:
- given a source shape M and a target shape N, pairing the vertices with nearest-neighbors
- obtaining the L2 loss and the direction in which each vertex should move to fit the surface
- computing the gradient of ARAP energy given by the deformed shape M w.r.t. the initial M 
- Taking an optimization step considering the two gradients, updating vertices' positions 
- iterating

I know there are some python implementations for ARAP in other libraries, like: https://libigl.github.io/tutorial/#as-rigid-as-possible
Another nice library to work with 3D meshes (which also has an ARAP implementation in python) is: http://www.open3d.org/html/tutorial/Advanced/mesh_deformation.html

However, I think it depends on your applicative case; if you need, for example, to backpropagate through the ARAP regularization, probably you can find some pytorch implementation here: https://github.com/GitBoSun/ARAPReg

Finally, if I remember correctly, ARAP energy is not particularly insane to implement; if I can share my two cents, probably considering a re-implementation of the original paper could be not so tricky, and you can obtain the best fit to what you need :)

I hope this answers your question, and let me know further doubts or comments!

Best,

Riccardo

**Comment by GiBucci** [comment]
Thank you for your prompt reply!

The process is clear, thank you.

I already saw the open3d implementation, but it was not clear how to use that since it is done in one block (meaning that iterations are done under the hood), and this is what I meant with different approach, since no one has a step-by-step gradient function.
The direct transformation is good but brings up some problems with respect to armpits, hands and crotch. Fingers disappear and armpits are "fused" together.

Can you give me a hint on what could be the problem? 
I suspect it is a knn error, it can be ambiguos to understand neighbors in close parts like fingers. Is it the reason why you do it step by step with a coefficient on the gradient? Or am I missing something that gives "rigidity" to the translation?

If that is the problem, I will try to implement my own version of the ARAP gradient. I am fairly new to geometry world and this is why I was trying to find something at least close to what I am tryng to achieve. But I think your answers can be already a great help for me.

Thank you again and have a nice day.

**Comment by riccardomarin** [comment]
Sorry, I missed the notification about your message.

I see that many libraries obfuscate the computations inside the code, and apart from GPToolBox, I do not know other libraries that provide an explicit function.

Yes, in our approach, after some steps (let's say T steps), we update KNN, and we recompute the ARAP energy considering v_T the "initial" one. This allows us to keep the regularization of ARAP strong (avoiding artefacts and collapsing) and not constrain ourselves to the initial guess that may be pretty far, especially for local things like fingers. We have some cases where, even if the template is not allowed to move fingers, the ARAP registration succeed in catching finger bending that is not too extreme (i.e., if the hand is a fist, I think there is no chance with this approach).

One option could also be to implement the ARAP energy and let some auto-differentiating library does the "dirty" job :) It would be interesting to me; if you decide to follow this road, let me know, and I may be able to support you to some extent.

Good luck and have a nice day!

**Comment by GiBucci** [comment]
Don't worry.

At the moment I used mostly libigl blocks to reimplement ARAP energy in a similar fashion.
The only problem is at some point the results are different and my matrices reach infinite or nan values, so single value decomposition doesn't converge. I am clearly doing something wrong. Do you remember if you met this problem somewhere and had to add some tricks to avoid this kind of result?
I'll see if I can fix something or I'll try your way and reimplement each basic function.

Regarding this approach, I can tell you it is clear now, but I feel that probably in my case it won't be able to model fingers properly anyway, but it's not an approach problem, but an input data issue.

Finally, my goal right now is to have something working as soon as possible, so my main focus will be to have a "replica" of the basic approach. When I'll make it work, I can take some time to study auto differentiation. As I said I am not very expert in this field, so I'll need some time to study what you mean and the approach.
If I decide to pursue this second road, I'll contact you! I don't know if you have any direct contact, or I'll write you here in case.

In the meantime, thanks again for your kindness, and have a nice day.

**Comment by riccardomarin** [comment]
Probably it could happen for some pathological cases, like collapsed vertices or numerical instability. I remember I had similar problems; in some specific situations, the ARAP energy may promote some "spikes" over the surface that diverge during the optimization (that's also why we proceed with small steps in the optimization). I expect this happens in specific configurations that you may spot by manual investigation. I am also aware that the original ARAP implementation has some trouble since the cotangent weights may provide negative weights for low-quality meshes (you can find further details and some fixes like the Spokes and Rims here: https://doc.cgal.org/latest/Surface_mesh_deformation/index.html )

Sure, feel free to contact me at: marin@di.uniroma1.it

Best!

**Comment by GiBucci** [comment]
No sorry I didn't explain myself. My python implementation is very similar to yours, but mine "explodes" after around 20 iterations, given same input and same output. Your code works from start to end so it is not a problem related to the case.
I am probably missing something, but it seems that matlab and pyhton do different calculations and it's hard for me to spot the error.

Anyway, I'll keep trying. 
Thank you for your mail address. 
I'll try not to bother you at least for this first step, I promise!

Tell me if you prefer to close the issue, or it is fine to keep it open for inquiries.
